geopandas/tools/sjoin.py

```diff
@@ -275,59 +276,76 @@ def sjoin_nearest(
+            # warn about using rtree < 0.9.4 and max_search_neighbors option
+            # see https://github.com/Toblerity/rtree/pull/141
+            raise ValueError("max_search_neighbors must be an integer and >= 1")
 
-        if [int(i) for i in RTREE_VERSION.split(".")] < [0, 9, 4]:
+        if str(RTREE_VERSION) < LooseVersion("0.9.4"):
             warn(
                 "Using an rtree version < 0.9.4 may cause inconsistent "
                 "results when using max_search_neighbors. Consider using a "
                 "large number or max_search_neighbors=None."
             )
-    if search_radius is not None:
-        assert search_radius >= 0, "search_radius must be >=0"
+    if search_radius is not None and search_radius < 0:
+        raise ValueError("search_radius must be >= 0")
 
-    def query_index(geo_in_l):
+    def query_index(geo_in_l, buff_geo_in_l):
         """
         Queries the spatial index to filter results.
         """
         # restric by radius
         if search_radius is not None:
-            if geo_in_l.geom_type == "Point" and search_radius == 0:
-                # applying .buffer(0) to a point returns an empty geometry
-                geo_search = geo_in_l.coords[0]
-            else:  # polygon or buffer > 0
-                geo_search = geo_in_l.buffer(search_radius).bounds
-            in_radius = list(tree_idx.intersection(geo_search))
+            in_radius = list(tree_idx.intersection(buff_geo_in_l))
+            check_radius = True
         else:
             in_radius = right_df.index
+            check_radius = False
         # find neighbors
         if max_search_neighbors is not None:
             neighbors = list(
                 tree_idx.nearest(geo_in_l.bounds, num_results=max_search_neighbors)
             )
+            check_neighbors = True
         else:
             neighbors = right_df.index
-        # ensure indexes are int
-        return (int(idx) for idx in neighbors if idx in in_radius)
+            check_neighbors = False
+
+        if check_radius and check_neighbors:
+            return np.array(
+                [int(idx) for idx in neighbors if idx in in_radius], dtype=int
+            )
+        elif check_radius:
+            return np.array(in_radius, dtype=int)
+        elif check_neighbors:
+            return np.array(neighbors, dtype=int)
+        else:
+            return right_df.index  # check all indexes in right_df
 
     # the accuracy of the spatial index is limited, so we need to manually
     # check each set of matches for actual distance
     l_idx = []
     r_idx = []
     distances = []
-    for ind_in_left, geo_in_l in enumerate(left_df.geometry):
+
+    # pre-buffer the bounds of right_df geometries by search radius
+    # these will be used by the spatial index
+    s_r = search_radius or 0
+    bbox_delta = np.array([-s_r, -s_r, s_r, s_r])  # minx, miny, maxx, maxy
+    buff_geos_in_l = (left_df.geometry.bounds + bbox_delta).values
```

Comment:
Can we skip this if `search_radius` is 0 or None. There's no need for this operation then.

Reply:
done